### ISSUE TITLE
How can I set headers for resources in the public folder?

### ISSUE THREAD

**Issue by dataexcess** [nuxt/nuxt#22141]
### Environment

------------------------------
- Operating System: Darwin
- Node Version:     v18.16.1
- Nuxt Version:     3.6.2
- Nitro Version:    2.5.2
- Package Manager:  npm@9.8.0
- Builder:          vite
- User Config:      devtools, modules, css, app, routeRules, security, vite
- Runtime Modules:  @nuxtjs/tailwindcss@6.8.0, @pinia/nuxt@0.4.11, nuxt-security@0.13.1
- Build Modules:    -
------------------------------

### Reproduction

/

### Describe the bug

Hello!

I have a nuxt 3 app and when a certain 'boot' button is clicked a synth.worker.js file is loaded from my nuxt public folder. However, the file is not being loaded, in google chrome dev tools it says the file is being blocked by Cross-Origin-Embedder-Policy not being set and should be set to 'require-corp'.

I tried to use nuxt-security to add the Cross-Origin-Embedder-Policy and all the others, but it is still being blocked. I also tried to run some server/middleware file but it is not even triggered when the resource is dynamically requested.

How can I simply specifiy headers for static files in the public folder in nuxt 3????

Said file is located in /public/lib/scserver/scsynth.worker.js

This is my nuxt.config.ts:
// https://nuxt.com/docs/api/configuration/nuxt-config
```
export default defineNuxtConfig({
    devtools: { enabled: true },
    modules: ['@nuxtjs/tailwindcss', '@pinia/nuxt', 'nuxt-security'],
    css: ['@/assets/css/main.css'],
    app: {
        head: {
            title: 'ColliderNFT',
            htmlAttrs: {
                lang: 'en'
            },
            script: [
                {
                    src: 'lib/scserver/scsynth.js',
                    body: true,
                    crossorigin: 'anonymous'
                }
            ]
        }
    },
    routeRules: {
        '/lib': { cors: true }
    },
    security: {
        headers: {
            crossOriginOpenerPolicy: 'same-origin',
            crossOriginResourcePolicy: 'cross-origin',
            crossOriginEmbedderPolicy: 'require-corp',
            contentSecurityPolicy: { 'frame-ancestors': ["'self'", 'https://localhost:3000'] },
            strictTransportSecurity: 'max-age=31536000; includeSubDomains'
        }
    },
    vite: {
        server: {
            headers: {
                'Access-Control-Allow-Origin': '*',
                'Cross-Origin-Embedder-Policy': 'require-corp'
            },
            cors: {
                origin: 'http://localhost:3000'
            }
        }
    }
})
```

**Comment by danielroe** [comment]
You can set headers using `routeRules`. Note that `vite.server` options are misnamed and only apply to the vite _dev server_ which is not used by Nuxt.

```ts
export default defineNuxtConfig({
  routeRules: {
    '/lib/scserver/scsynth.worker.js': {
      headers: {
        // your headers
      }
    }
  },
})
```

**Comment by dataexcess** [comment]
> You can set headers using `routeRules`. Note that `vite.server` options are misnamed and only apply to the vite _dev server_ which is not used by Nuxt.
> 
> ```ts
> export default defineNuxtConfig({
>   routeRules: {
>     '/lib/scserver/scsynth.worker.js': {
>       headers: {
>         // your headers
>       }
>     }
>   },
> })
> ```

Thank you for the help. I tried your code but it is not working. However I noticed that in the network tab, the Initiator is 'other', maybe this means that this file is requested by another file, probably scsynth.js and not the main nuxt app? Could this be a reason that it is not going through the nuxt routes?

<img width="1429" alt="Screenshot 2023-07-14 at 16 25 02" src="https://github.com/nuxt/nuxt/assets/51317856/398fc21f-4b41-4480-a3ad-d3210fad0d4b">

**Comment by dataexcess** [comment]
I managed to circumvent this issue by copy/pasting the code of the worker.js file inside my calling javascript file and simply using this:

```
const workerCode = `pasted worker.js file code here...`
const blob = new Blob([workerCode], { type: 'text/javascript' });
const workerUrl = URL.createObjectURL(blob);
```

All good. Thnx

**Comment by laurensV** [comment]
Not really all good, this is just a hack around this issue.. We should be able to set headers for static files.. (I have the same problem with a worker file)

**Comment by kogratte** [comment]
Guys, also stucked with emscripten files not served with the expected headers.

Middleware are not designed to do the job.
Plugins are inefficients, as we're speaking about static assets.
RouteRules doesn't works either.

Nitro config is quiet.
Nuxt config is silent.

It sucks, really. How are we supposed to deal with those kind of requirements?

**Comment by carbdias** [comment]
Is it possible to do this in route middleware ?
I need to apply a locale variable in the header, which is not available in the nuxt.config file.

**Comment by queroin** [comment]
Did some digging - by reading this [post](https://github.com/nuxt/devtools/issues/463#issuecomment-1781967728) I finally get it. (Thanks to [AlanCodeLog](https://github.com/AlansCodeLog))

> you can't add headers through nuxt to public resources when using the dev server. You must set the headers using vite

Therefore, below is my vite config

```
export default defineConfig({
...
  server: {
    headers: {
      'cache-control': 'max-age=31536000'
    }
  }
});
```

run `nuxt build` (rebuild your `outDir`) then use `npm start` NOT `npm run dev`

@danielroe - Is this documented somewhere in the nuxt 3 docs? I can't seem to find any notes saying it will not take effect for `outDir` or `public` folders, maybe you could help me put a link here. Thanks!

**Comment by queroin** [comment]
Deleted my first post here's what works for me.

Did some digging - by reading this [post](https://github.com/nuxt/devtools/issues/463#issuecomment-1781967728) I finally got a big lead. (Thanks to [AlanCodeLog](https://github.com/AlansCodeLog))

> you can't add headers through nuxt to public resources when using the dev server. You must set the headers using vite

Therefore, below is my vite config

```
export default defineConfig({
...
  server: {
    headers: {
      'cache-control': 'max-age=31536000'
    }
  }
});
```

But it doesn't take effect when u add it in the vite config, tried a few tests, and I'm back to using the routeRules - you just need to specify the proper route to your public files. Here's my nuxt config file

```
export default defineNuxtConfig({
 ...
  routeRules: {
    '/my-js-file.umd.js': {
      'headers': {
        'Cache-Control': 'public, no-cache, weszdasfwe'
      }
    }
  }
});
```
^ browser URL sample `localhost:3000/my-js-file.umd.js`

run `nuxt build` (rebuild your outDir) then use `npm start` NOT `npm run dev`. Run in Incognito for optimal results.

@danielroe - Is this documented somewhere in the nuxt 3 docs? I can't seem to find any notes saying it will not take effect for outDir or public folders, maybe you could help me put a link here. Thanks!

**Comment by danielroe** [comment]
You should not have a `vite.config` if you are using Nuxt. You can set vite config within your nuxt config `vite` option.

But changing anything related to `server` will _only_ impact your dev server. So`routeRules` or nitro configuration or middleware is the correct approach.

**Comment by AlansCodeLog** [comment]
@queroin Sorry maybe my comment wasn't clear. This must be done with a vite plugin that adds a middleware, for example, I use something like this, you can customize this further for different routes, etc:

```ts
// in some file somewhere (e.g. build/plugins/viteHeadersPlugin)
export function viteHeadersPlugin(headers: Record<string, string>): Plugin {
	return {
		name: "configure-response-headers",
		configureServer: server => {
			server.middlewares.use((_req, res, next) => {
				for (const key of Object.keys(headers)) {
					res.setHeader(key, headers[key])
				}
				next()
			})
		},
	}
}
// in nuxt.config.ts
//...
	vite: {
		plugins: [
			// you can also inline the plugin
			viteHeadersPlugin({
				"Access-Control-Allow-Origin":"*"
				// ...any other headers
			})
		]
	}
//...
```
`vite.server.headers` is ignored, I'm not 100% sure why.

@danielroe None of the recommended options work in the dev server. Both a recommended way + this must be used to get consistent behavior between dev and production. 

Is routeRules supposed to already do this already when using the dev server but it doesn't? Because that would be ideal. 

Also in a comment you wrote nuxt does not use the vite dev server... do you mean the vite server options? Is that why `vite.server.headers` is ignored? If that's on purpose should the types be changed to remove that property?

**Comment by Niki2k1** [comment]
Setting headers for nitro public assets in production works just fine with this:
```js
export default defineNuxtConfig({
  nitro: {
    routeRules: {
      '/my-route/**': {
        headers: {
          'cross-origin-embedder-policy': 'require-corp',
        },
      },
    },
  }
}
```

But in dev the header does not get set.

**Comment by xak2000** [comment]
@danielroe 
> So `routeRules` or nitro configuration or middleware is the correct approach.

Is it **intended** that `routeRules` or nitro configuration doesn't work in DEV?
Maybe it's just a bug and should be reported?

If it is intended, then how could you call it "correct approach"? How to develop the app if headers will be set only in PROD? Other parts of the app could not work properly if for example they expect `application/json` but receive `application/octet-stream`...

I mean, these `routeRules` are important part of the configuration and should work identically in both DEV and PROD, or if this is not possible then their usage should be discouraged.

**Comment by xak2000** [comment]
Should I open a bug for this, then? Because they definitely do not work in DEV mode in the same way as in PROD mode, at least with static files.

In [this example](https://stackblitz.com/edit/nuxt-route-rules-dev?file=nuxt.config.ts) `/hello.txt` route should return `Content-type: application/json` header, and it **does work** in the PROD mode, but **doesn't work** in the DEV mode, where `Content-type: text/plain; charset=UTF-8` is returned.

**Comment by kogratte** [comment]
An easy manner to know if the behaviour is good enough, is to reply to the
quick question : are we able to serve wasm using dev / prod server? As it
requires specific headers on assembly file, the answer should be yes,
otherwise the issue shouldn't be closed.

**Comment by danielroe** [comment]
Please open a new issue with a reproduction if you are encountering an issue 🙏

**Comment by dataexcess** [comment]
Hey hey I'm revisiting the website that started this issue and would really prefer to load the external file as opposed to my hack. 
Following issue [26740](https://github.com/nuxt/nuxt/issues/26740) I thought this was a nitro issue which they fixed?
In my case I still get the same error... this is my nuxt config:

```
routeRules: {
		'/scsynth.worker.js': {
			headers: {
				'Content-Type': 'application/javascript',
				'Cross-Origin-Opener-Policy': 'same-origin',
				'Cross-Origin-Embedder-Policy': 'require-corp',
				'Cross-Origin-Resource-Policy': 'same-origin'
			}
		}
	},
	nitro: {
		routeRules: {
			'/scsynth.worker.js': {
				headers: {
					'Content-Type': 'application/javascript',
					'Cross-Origin-Opener-Policy': 'same-origin',
					'Cross-Origin-Embedder-Policy': 'require-corp',
					'Cross-Origin-Resource-Policy': 'same-origin'
				}
			}
		}
	}
```

Any ideas?

**Comment by dataexcess** [comment]
Ow seems they're looking at it but it's still not merged? so there's no way to do this currently.
